New Onboarding: Use Redux for users in `EditorCheckoutModal`

tyxla opened the pull request:
#### Changes proposed in this Pull Request

In `EditorCheckoutModal` we use `lib/user` to check whether the user is logged in. That is necessary to build a proper cart key. This PR refactors this usage to use the Redux version of the current user instead. 

However, I was unable to reproduce a use case that displays `EditorCheckoutModal` for a logged-out user. Seems like the logged-out scenario isn't testable unless I'm missing something. 

This PR is part of #24004 where we aim to reduxify `lib/user`.

#### Testing instructions

* Go to `/new`
* Input a site title.
* Pick a paid domain.
* Pick some theme and font on the following steps.
* Stick with the "Custom domains" on the features step.
* Pick the "Free" plan on the plans step
* When you're redirected to the editor, click "Launch"
* Select your paid domain again.
* Pick a paid plan.
* Click "Launch your site"
* You should now see the editor checkout modal.
* Verify it works alright by completing the checkout process and looking out for errors.

tyxla committed:
Gutenboarding: Use Redux for users in EditorCheckoutModal

jsnajdr commented:
> However, I was unable to reproduce a use case that displays EditorCheckoutModal for a logged-out user.

Looking at how the logged-in check was added in #46860, it seems to me that `isLoggedOutCart` is simply a required parameter of `getCartKey`. So the component looks at the true value of that, although in practice we're always logged in.

jsnajdr commented on client/blocks/editor-checkout-modal/index.tsx:
```diff
@@ -2,7 +2,7 @@
  */
 import React, { useMemo, useEffect } from 'react';
-import { connect } from 'react-redux';
+import { connect, useSelector } from 'react-redux';
```
Let's use just one way of using `react-redux`: `connect` HOC or the hooks. The component is already functional and removing `connect` should be very easy.

tyxla replied:
Good point, addressed in https://github.com/Automattic/wp-calypso/pull/53184/commits/b54e7b66ab4fcaed94f82616b599ec9d0760ee90

tyxla committed:
Remove connect() and use useSelector instead

tyxla committed:
Use the locale slug from useTranslate

tyxla commented:
> The same info is also on `translate.localeSlug` (from `useTranslate`). And it's closer to being the "source of truth" regarding the current locale.

Good call, also TIL we're exposing the locale slug there. I've addressed this in 369ebf0358de769848cf2200901052dc13071bb9. Had to update the types, because they were not aware of the locale slug being exposed there.

> Let's not add new usages of `redux-bridge`. If `canUserPurchageGSuite` needs some information about the user, it should get it as function parameter.

I was thinking about that, however I reconsidered. Disregarding areas that can get access to the Redux store from the context, we have this tree of functions depending on each other that would have to all be updated to pass the user data as an argument:

* `canUserPurchaseGSuite`
	* `canDomainAddGSuite`
		* `getEligibleGSuiteDomain`
		* `getGSuiteSupportedDomains`
			* `hasGSuiteSupportedDomain`

I'm not convinced this is not an overkill, considering that these functions are not necessarily directly related to the user in any way. 

What do you think?

tyxla committed:
Fix mock

jsnajdr commented:
> we have this tree of functions depending on each other that would have to all be updated to pass the user data as an argument

I still think it's totally worth it to make the dependency on user explicit. And we don't need to update all these functions. Instead, we can remove the `canUserPurchaseGSuite` check from `canDomainAddGSuite` and make these functions answer questions only about the domains, nothing else.

It's then the calling code's responsibility to call `canUserPurchaseGSuite` explicitly. For example, the Home screen shows an "Add Email" quick link if the site has an eligible domain, currently checked with
```js
hasCustomDomain = getGSuiteSupportedDomains( domains ).length > 0
```
There would need to be an additional `canUserPurchaseGSuite` check here, easy to implement as Redux selector.

I believe that's a much easier patch, without any domino effects.

tyxla committed:
Refactor canUserPurchaseGSuite to a selector and use throughout Calypso

tyxla commented:
> > we have this tree of functions depending on each other that would have to all be updated to pass the user data as an argument
> 
> I still think it's totally worth it to make the dependency on user explicit. And we don't need to update all these functions. Instead, we can remove the `canUserPurchaseGSuite` check from `canDomainAddGSuite` and make these functions answer questions only about the domains, nothing else.
> 
> It's then the calling code's responsibility to call `canUserPurchaseGSuite` explicitly. For example, the Home screen shows an "Add Email" quick link if the site has an eligible domain, currently checked with
> 
> ```js
> hasCustomDomain = getGSuiteSupportedDomains( domains ).length > 0
> ```
> 
> There would need to be an additional `canUserPurchaseGSuite` check here, easy to implement as Redux selector.
> 
> I believe that's a much easier patch, without any domino effects.

Thanks for taking the time to dig through this, that makes sense. Although it's a larger refactor than I had hoped it would be, it's a good one because it makes the logic more obvious and explicit, and that's always a good reason to spend some extra time.

In https://github.com/Automattic/wp-calypso/pull/53184/commits/7757fc076c1ce1b2f41c5cc077a40960c08867a8 I've refactored `canUserPurchaseGSuite` to a selector and refactored all code that was relying on the previous helper to call that selector. 

Please take a look and let me know what you think, @jsnajdr.

I'd also love it if folks from @Automattic/letero could take a look and verify we're not breaking anything here.

daledupreez commented:
I'll take a look for Letero tomorrow morning UTC - I won't get to this today.

tyxla commented:
Thanks a bunch, @daledupreez

daledupreez commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -82,7 +84,9 @@ class EmailManagementHome extends React.Component {
-						isGSuiteSupported={ hasGSuiteSupportedDomain( [ selectedDomain ] ) }
+						isGSuiteSupported={
+							hasGSuiteSupportedDomain( [ selectedDomain ] ) && userCanPurchaseGSuite
+						}
```
Nite: Here and later on, does it make sense to check the local boolean first?
```suggestion
						isGSuiteSupported={
							userCanPurchaseGSuite && hasGSuiteSupportedDomain( [ selectedDomain ] ) 
						}
```

tyxla replied:
Sure thing 👍  Reordered the conditions in 4e7e388966aa0848a18f7287310dc587c4a1b853

daledupreez commented on client/my-sites/customer-home/cards/actions/quick-links/index.jsx:
```diff
@@ -297,7 +298,8 @@ const mapStateToProps = ( state ) => {
 		menusUrl: getCustomizerUrl( state, siteId, 'menus' ),
 		isNewlyCreatedSite: isNewSite( state, siteId ),
 		showCustomizer: ! isSiteUsingFullSiteEditing( state, siteId ),
-		hasCustomDomain: getGSuiteSupportedDomains( domains ).length > 0,
+		hasCustomDomain:
+			getGSuiteSupportedDomains( domains ).length > 0 && canUserPurchaseGSuite( state ),
```
Letero will need to improve this logic further, as there are more cases where can promote Professional Email (which has less restrictive rules compared to G Suite.)

tyxla replied:
👍  Let me know if I can do anything to improve that in this PR.

daledupreez replied:
I think we're good for now -- this needs a business logic change rather than a refactor! Let's get this shipped and we can take on working out what the business logic ought to be separately.

tyxla replied:
Thanks for verifying, @daledupreez. Feel free to ping me in any follow-ups if I can help with anything.

daledupreez reviewed (commented):
I haven't had a chance to test this yet, but the code is looking good. I'll have a chance to wrap  the testing later today.

If anyone has some time to kick the tires, the main flows that need to be tested are as follows:
### Flow 1

 * From within Calypso, go to Upgrades -> Domains.
 * Click on the "Add a domain" button in the top right corner, and then click on "to this site" in the popover.
 * Search for a domain and select it (we're not going to check out, so don't worry about what you pick)
 * Confirm that you see a Google Workspace upsell and that you can interact with it

### Flow 2

* Go to Upgrades -> Emails for a site with one or more domains without a paid email subscription (and which hasn't had a G Suite subscription in the past)
* Click on "Add Email" for such a domain at the bottom of the screen
* Verify that you see a card for Google Workspace in the email comparison page

There are then two more subtle edge cases for the new email home page, but I'll test those tonight.

tyxla committed:
Reorder conditions

tyxla commented:
Thanks for giving it a thorough look, @daledupreez! I'll be standing by for your final feedback once you get to it.

daledupreez reviewed (approved):
This LGTM!

For completeness, the two edge cases [I mentioned previously](https://github.com/Automattic/wp-calypso/pull/53184#pullrequestreview-671090823) are as follows:

 * The user is visiting `/email/:site_slug` where `:site_slug` is a domain that doesn't have email and it's the only domain on the site (this is the case for a domain with a single domain that is set to primary)
 * The user is visiting `/email/:site_slug` where `:site_slug` is a `.wordpress.com`, `*.*.blog`, or `.wpcomstaging.com` subdomain and the user has exactly one domain on their site that is not the primary domain and doesn't have an email subscription. (This can be set up by making the free subdomain for a site primary when that site has one actual domain on it.

tyxla commented:
Thanks, @jsnajdr, and @daledupreez for the thorough reviews 🙌
